File: services/api/src/main/java/com/launchly/release/services/ReleaseService.java
package com.launchly.release.services;

import com.launchly.audit.enums.AuditAction;
import com.launchly.audit.services.AuditService;
import com.launchly.project.entities.Project;
import com.launchly.project.repositories.ProjectRepository;
import com.launchly.release.dto.GateExemptionRequest;
import com.launchly.release.dto.ReleaseRequest;
import com.launchly.release.dto.ReleaseResponse;
import com.launchly.release.entities.GateExemption;
import com.launchly.release.entities.Release;
import com.launchly.release.enums.ReleaseStatus;
import com.launchly.release.repositories.GateExemptionRepository;
import com.launchly.release.repositories.ReleaseRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class ReleaseService {
    private final ReleaseRepository releaseRepository;
    private final GateExemptionRepository exemptionRepository;
    private final GateCheckService gateCheckService;
    private final ProjectRepository projectRepository;
    private final AuditService auditService;

    public ReleaseService(ReleaseRepository releaseRepository,
                          GateExemptionRepository exemptionRepository,
                          GateCheckService gateCheckService,
                          ProjectRepository projectRepository,
                          AuditService auditService) {
        this.releaseRepository = releaseRepository;
        this.exemptionRepository = exemptionRepository;
        this.gateCheckService = gateCheckService;
        this.projectRepository = projectRepository;
        this.auditService = auditService;
    }

    @Transactional
    public ReleaseResponse createRelease(String projectId, ReleaseRequest request, String userId) {
        Release release = new Release();
        release.setProjectId(projectId);
        release.setEnvironmentId(request.environmentId());
        release.setDeploymentId(request.deploymentId());
        release.setVersion(request.version() != null ? request.version() : generateVersion());
        release.setNotes(request.notes());
        release.setStatus(ReleaseStatus.DRAFT);

        release = releaseRepository.save(release);

        // Auto-check gates
        checkAndUpdateGates(release);

        return ReleaseResponse.from(release);
    }

    public List<ReleaseResponse> listReleases(String projectId) {
        return releaseRepository.findByProjectIdOrderByCreatedAtDesc(projectId)
                .stream().map(ReleaseResponse::from).collect(Collectors.toList());
    }

    public ReleaseResponse getRelease(String id) {
        Release r = releaseRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Release not found: " + id));
        return ReleaseResponse.from(r);
    }

    public GateCheckResult getGateStatus(String id) {
        Release r = releaseRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Release not found: " + id));
        return gateCheckService.checkGates(r.getProjectId(), r.getEnvironmentId(), r.getDeploymentId());
    }

    @Transactional
    public ReleaseResponse publish(String id, String userId) {
        Release r = releaseRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Release not found: " + id));

        // Check gates before publishing
        GateCheckResult gates = gateCheckService.checkGates(
            r.getProjectId(), r.getEnvironmentId(), r.getDeploymentId());

        // Count exemptions
        List<GateExemption> exemptions = exemptionRepository.findByReleaseId(r.getId());
        boolean allGatesPassed = gates.getResults().stream()
            .allMatch(g -> g.isPassed() || exemptions.stream().anyMatch(e -> e.getGateName().equals(g.getGateName())));

        if (!allGatesPassed) {
            throw new IllegalStateException("Not all gates passed or exempted. Cannot publish.");
        }

        r.setStatus(ReleaseStatus.PUBLISHED);
        r.setGateStatus("PASSED");
        r.setReleasedBy(userId);
        r.setReleasedAt(Instant.now());
        Release saved = releaseRepository.save(r);
        auditService.record(userId, workspaceId(saved.getProjectId()), AuditAction.PUBLISH_PRODUCTION, "release", saved.getId(),
                java.util.Map.of("projectId", saved.getProjectId(), "version", saved.getVersion()));
        return ReleaseResponse.from(saved);
    }

    @Transactional
    public GateExemption exemptGate(String releaseId, String gateName, GateExemptionRequest request, String userId) {
        Release r = releaseRepository.findById(releaseId)
                .orElseThrow(() -> new IllegalArgumentException("Release not found: " + releaseId));

        GateExemption exemption = new GateExemption();
        exemption.setReleaseId(releaseId);
        exemption.setGateName(gateName);
        exemption.setExemptedBy(userId);
        exemption.setReason(request.reason());
        exemption = exemptionRepository.save(exemption);
        auditService.record(userId, workspaceId(r.getProjectId()), AuditAction.GATE_EXEMPT, "release", r.getId(),
                java.util.Map.of("gateName", gateName));

        // Re-check gates
        checkAndUpdateGates(r);

        return exemption;
    }

    public List<GateExemption> getExemptions(String releaseId) {
        return exemptionRepository.findByReleaseId(releaseId);
    }

    private void checkAndUpdateGates(Release release) {
        GateCheckResult result = gateCheckService.checkGates(
            release.getProjectId(), release.getEnvironmentId(), release.getDeploymentId());

        List<GateExemption> exemptions = exemptionRepository.findByReleaseId(release.getId());
        boolean allPassed = result.getResults().stream()
            .allMatch(g -> g.isPassed() || exemptions.stream().anyMatch(e -> e.getGateName().equals(g.getGateName())));

        if (allPassed) {
            release.setStatus(ReleaseStatus.READY);
            release.setGateStatus("PASSED");
        } else {
            release.setStatus(ReleaseStatus.PENDING_GATES);
            release.setGateStatus("FAILED");
        }
        releaseRepository.save(release);
    }

    private String generateVersion() {
        java.time.LocalDate now = java.time.LocalDate.now();
        return "v" + now.getYear() + "." + now.getMonthValue() + "." +
               String.format("%03d", releaseRepository.count() + 1);
    }

    private String workspaceId(String projectId) {
        return projectRepository.findById(projectId).map(Project::getWorkspaceId).orElse(null);
    }
}

File: services/api/src/release/release.service.ts
import { BadRequestException, ForbiddenException, Injectable, NotFoundException, Optional } from '@nestjs/common';
import { PrismaService } from '../common/prisma/prisma.service';
import { GateCheckService } from './gate-check.service';
import { AuditService } from '../audit/audit.service';

/**
 * 发布（Release）Service（KI-020 / KI-021 / R0-09 / R5）。
 *
 * 关键约束：
 * - publish 必须有完整的 Gate 评估结果；空 gates + allPassed=false 视为配置异常，fail closed。
 * - 同 digest 晋级不重新构建；gate status 写入用 PASSED/EXEMPTED/FAILED 三态明确区分。
 */

@Injectable()
export class ReleaseService {
  constructor(
    private readonly prisma: PrismaService,
    private readonly gateCheck: GateCheckService,
    @Optional() private readonly audit?: AuditService,
  ) {}

  async createRelease(projectId: string, data: { environmentId: string; deploymentId: string; version: string; notes?: string }, userId: string) {
    if (!data.version || !data.version.trim()) {
      throw new BadRequestException('version 不能为空');
    }
    const environment = await this.prisma.environment.findUnique({
      where: { id: data.environmentId },
      select: { projectId: true },
    });
    if (!environment) throw new BadRequestException('环境不存在');
    if (environment.projectId !== projectId) throw new BadRequestException('环境不属于当前项目');

    const deployment = await this.prisma.deployment.findUnique({
      where: { id: data.deploymentId },
      select: { projectId: true },
    });
    if (!deployment) throw new BadRequestException('部署不存在');
    if (deployment.projectId !== projectId) throw new BadRequestException('部署不属于当前项目');

    return this.prisma.release.create({
      data: {
        projectId,
        environmentId: data.environmentId,
        deploymentId: data.deploymentId,
        version: data.version,
        notes: data.notes ?? null,
        releasedBy: userId,
      },
    });
  }

  async listReleases(projectId: string) {
    return this.prisma.release.findMany({ where: { projectId }, orderBy: { createdAt: 'desc' } });
  }

  async getRelease(id: string) {
    const release = await this.prisma.release.findUnique({ where: { id } });
    if (!release) throw new NotFoundException('发布不存在');
    return release;
  }

  async getGateStatus(id: string) {
    return this.gateCheck.checkGates(id);
  }

  /**
   * 发布晋级。
   * KI-020 修复要点：
   * - 拒绝空 gates + allPassed=false 的"伪装 EXEMPTED"。
   * - 必须存在且通过（或被显式豁免）所有 Gate 才能发布。
   */
  async publish(id: string, userId: string) {
    const release = await this.prisma.release.findUnique({ where: { id } });
    if (!release) throw new NotFoundException('发布不存在');

    const gateStatus = await this.gateCheck.checkGates(id);

    // KI-020: 若 Gate 评估异常（结构矛盾/空集合 + allPassed=false），直接拒绝。
    if (!gateStatus || !Array.isArray(gateStatus.gates) || gateStatus.gates.length === 0) {
      throw new ForbiddenException('Gate 评估结果为空，无法判断发布安全，请先修复 Gate 配置');
    }
    if (gateStatus.gates.length === 0 && gateStatus.allPassed === false) {
      throw new ForbiddenException('Gate 评估结果矛盾（空集合且 allPassed=false），已 fail closed');
    }

    const exemptions = await this.prisma.gateExemption.findMany({ where: { releaseId: id } });
    const exemptedGates = new Set(exemptions.map(e => e.gateName));

    const unresolvedFailures = gateStatus.gates.filter(g => !g.passed && !exemptedGates.has(g.name));
    if (unresolvedFailures.length > 0) {
      throw new ForbiddenException('Gate 未通过: ' + unresolvedFailures.map(g => g.message).join('; '));
    }

    // EXEMPTED 状态判定：所有未通过的 gate 都已有豁免，且没有任何 unresolved failure。
    const hasUnresolvedFailure = unresolvedFailures.length > 0;
    return this.prisma.release.update({
      where: { id },
      data: {
        status: 'PUBLISHED',
        gateStatus: gateStatus.allPassed ? 'PASSED' : (hasUnresolvedFailure ? 'FAILED' : 'EXEMPTED'),
        releasedBy: userId,
        releasedAt: new Date(),
      },
    });
  }

  async exemptGate(id: string, gateName: string, data: { reason: string; ticket?: string }, userId: string) {
    if (!data.reason || !String(data.reason).trim()) {
      throw new BadRequestException('豁免原因不能为空（KI-020）');
    }
    const exemption = await this.prisma.gateExemption.create({
      data: {
        releaseId: id,
        gateName,
        exemptedBy: userId,
        reason: data.reason,
        ticket: data.ticket ?? null,
      },
    });
    if (this.audit) {
      const release = await this.prisma.release.findUnique({
        where: { id },
        select: { project: { select: { workspaceId: true } } },
      });
      await this.audit.record(userId, release?.project.workspaceId ?? null, 'RELEASE_GATE_EXEMPTED', 'RELEASE', id, {
        gateName,
        reason: data.reason,
        ticket: data.ticket ?? null,
      });
    }
    return exemption;
  }

  async getExemptions(id: string) {
    return this.prisma.gateExemption.findMany({ where: { releaseId: id } });
  }
}
